### ISSUE TITLE
Type inference message from Exercise 1.5 is ambigous

### ISSUE THREAD

**Issue by bboston7** [GaloisInc/cryptol#745]
Executing the example from Exercise 1.5 produces an ambiguous type inference message:

```
Cryptol> ((1, 2), False, (3-1, (4, True)))                                         
Showing a specific instance of polymorphic result:                                 
  * Using 'Integer' for type of 0th tuple field                                    
  * Using 'Integer' for type of 1st tuple field                                    
  * Using 'Integer' for type argument 'a' of '(Cryptol::-)'                        
  * Using 'Integer' for type of 0th tuple field                                    
((1, 2), False, (2, (4, True)))     
```

Which 0th and 1st fields are being referred to when tuples are nested like this?

Similarly, Exercise 1.6 produces warnings, but the field numbers are insufficient alone to deduce which elements the warnings apply to:

```                                                                                
Cryptol> ((1, 2), False, (3-1, (4, True))).2.1                                     
[warning] at <interactive>:1:18--1:21:                                             
  Defaulting type argument 'a' of '(-)' to [2]                                     
[warning] at <interactive>:1:2--1:8:                                            
  Defaulting type of 1st tuple field to [2]                                     
[warning] at <interactive>:1:2--1:8:                                            
  Defaulting type of 0th tuple field to [1]                                     
Showing a specific instance of polymorphic result:                              
  * Using 'Integer' for type of 0th tuple field                                 
(4, True)                                                                       
```

**Comment by brianhuffman** [comment]
We'll need to make a pass over the book to make sure everything works as expected before the next release.

To avoid problems like this in the future, we should probably extract all the code and exercises from the book to run as part of our regression test suite.

**Comment by yav** [comment]
Yeah, we should do this, although we should make sure that we are clear about the purposes of the exercises.   Looking at the context here, using a less polymorphic example might be better, though, I doubt the purpose was to show the actual warnings.

We probably should have some part of the book explaining the warnings though.  In particular, these examples show 2 different types of messages:
   1. The ones about "Showing a specific instance of a polymorphic result", and
   2. The warnings about "Defaulting type".

These are quite different, and serve different purposes.  In particular, (1) talks about the result of the computation, while (2) talks about your program.   Notice that in the case of (2), the field names can be combined with the locations of the warnings to identify what is being defaulted quite precisely.  We could probably do better in the case of (1), although this example is quite pathological

**Comment by robdockins** [comment]
These examples could perhaps be improved, but I think the real issue is that the "for type of nth tuple field", etc., clauses don't indicated precisely enough where the defaulting occurred.  It would probably be enough to add a source location in addition to the nice descriptions.
